### ISSUE TITLE
persistent-sqlite fails to load - ARM

### ISSUE THREAD

**Issue by bburdette** [yesodweb/persistent#356]
Error is as follows:

Loading package yesod-1.4.1.4 ... linking ... done.
Loading package haskell-lexer-1.0 ... linking ... done.
Loading package pretty-show-1.6.8 ... linking ... done.
Loading package persistent-sqlite-2.1.1.2 ... linking ... ghc: /home/bananapi/code/chordster/.cabal-sandbox/lib/arm-linux-ghc-7.8.2/persistent-sqlite-2.1.1.2/libHSpersistent-sqlite-2.1.1.2.a: unknown symbol `_GLOBAL_OFFSET_TABLE_'
ghc: unable to load package `persistent-sqlite-2.1.1.2'
cabal: Error: some packages failed to install:
chordster-0.0.0 failed during the building phase. The exception was:
ExitFailure 1
[bananapi@lemaker chordster]$

**Comment by bburdette** [comment]
persistent-sqlite-2.1.1 fails the same way.  package installs ok, but fails at the link step.

**Comment by bburdette** [comment]
The default ghc on arch comes with llvm 3.5.1 (IIRC), but ghc complains about it being too new.  Also it doesn't compile.  I'm using llvm 3.4.2 which seems to work on some other haskell programs.

[bananapi@lemaker chordster]$ llvm-diff --version
LLVM (http://llvm.org/):
  LLVM version 3.4.2
  Optimized build.
  Built Aug  5 2014 (20:54:41).
  Default target: armv7l-unknown-linux-gnueabihf
  Host CPU: (unknown)

**Comment by snoyberg** [comment]
I'm afraid I have no experience with ARM, nor any systems on which to test this.

**Comment by bburdette** [comment]
I have a system running with this code on it - you're welcome to log in remotely if you like.  

The good news is it gets pretty close to completing - its in the link step after all.  If I have any luck with it I'll post back.

**Comment by creichert** [comment]
@bburdette What linker options are you specifying? 

Have you tried anything like:
```
    cc-options:     -fPIC
```

Are you building persistent-sqlite directly or your own library with persistent-sqlite as a dependency?  Do you have sqlite3 installed?

**Comment by bburdette** [comment]
sqlite appears to be installed - according to pacman its sqlite 3.8.8.2-1.  I'm using persistent-sqlite in this [little yesod project](https://github.com/bburdette/chordster).  I haven't tried -fPIC, I'll enable that and report back.  Compile is pretty slow so it might be a while.  I haven't enabled any other compiler flags, just used the library as-is.

**Comment by bburdette** [comment]
TLDR: same result compiling with -fPIC (I think) and running cabal test.  

I got a little further with things - the vector library is a problem as 0.10.12.2 won't compile on arm.  vector 0.11.0.0 does compile and I built persistent-sqlite with that.  I used the 'systemlib' flag which should enable -fPIC, so it was like this:

cabal install -f systemlib

So I thought I'd just run cabal test for persistent-sqlite.  This required hspec and persistent-template, so I compiled those normally with cabal install.  Unfortunately compiling the test program fails with the same error, unknown symbol _GLOBAL_OFFSET_TABLE_.  

What's a mystery to me is how my chordster project got to the link step, seeing as vector 0.10.12.2 fails to compile.  Maybe cabal decided to use a slightly different version of vector that doesn't have the same problem.  Hard to tell now since I deleted the sandbox I was using for that.

**Comment by bburdette** [comment]
So I created a new sandbox for chordster and recompiled to see what version of vector is used.  Turns out 0.10.12.2, the same version that refuses to compile in the global cabal environment.  Also yesod-auth and wai-app-static failed to compile, but then on doing cabal install a second time they succeeded apparently, and the process went on to fail on the persistent-sqlite linkage again.

**Comment by meh9** [comment]
I am getting the same error when building something that depends on persistent. However, my configuration is different, I am on Intel and not on ARM, but I did some testing and think it may still be relevant.

I tested building the package on identical Vagrant instances of Ubuntu 14.04 - 32-bit and 64-bit. When building on 32-bit it always fails with this error, on 64-bit the build succeeds. (It also fails on a 14.04 32-bit laptop, which is how I found it. Also tested 14.10-32bit on Vagrant which also fails.)

I am assuming that you are building on 32-bit ARM?

**Comment by bburdette** [comment]
yep, 32 bit ARMv7.  Interesting that it might have something to do with 32 vs 64 bit.

**Comment by bburdette** [comment]
Looked at this again, and found a workaround!  Downloaded persistent and changed this systemlib flag to True in the persistent-sqlite.cabal.  Added persistent-sqlte to my sandbox with add-source, and it works.  I thought I had tried that, but maybe not.  Anyway this was with persisent-sqlite-2.1.3.  

```
 flag systemlib
  description: Use the system-wide sqlite library
  default: True
```

So now my project compiles and so far runs normally.
